File: script/get-comic.py
# -*- coding: UTF-8 -*-
import requests
from requests.adapters import HTTPAdapter
import json
import re
import os
import urllib.parse
import utils

retryReq = requests.Session()
retryReq.mount('http://', HTTPAdapter(max_retries=2))
retryReq.mount('https://', HTTPAdapter(max_retries=2))


baseRoot= '../comic/test/'
# baseRoot= '../comic//'
maxPageNum = 200
fileType = ".html"
comicMark = '933'
reqList="https://www..com/manhua/" + comicMark + "/"
chapterFile = baseRoot + 'chapter.json'
imagesJsonFileName = 'images.json'
emptyJsonFileName = 'empty.json'
totalJsonFileName = 'total.json'


# 下载图片
def down_img(localFold,data,isDirect):
  # 漫画网站本来应该有，但返回了空的数据
  emptyImageList = []
  fold_name = urllib.parse.unquote(localFold) + '/'
  emptyFile = baseRoot+fold_name+ emptyJsonFileName
  emptyRecord = []
  if (os.path.exists(emptyFile)):
    with open(emptyFile, "r",encoding="utf-8") as f:
        content = f.read()
        if (content):
          emptyRecord = json.loads(content)
  img_index = 0
  img_len = len(data)
  while img_index < img_len:
    img_item = data[img_index]
    if (not img_item):
      img_index += 1
      continue

    img_item_arr = img_item.split('/')
    img_arr_len = len(img_item_arr)
    img_item_arr_last = img_item_arr[img_arr_len-1]
    imgCount = img_index+1
    if (isDirect):
      filename = baseRoot +fold_name + str(imgCount) + utils.getImageType(img_item)

    if (os.path.exists(filename)):
      print('chapter '+ localFold + '-' +str(img_index)+' exists')
    else:
      res = retryReq.get(img_item,timeout=60)
      # print(res.status_code)
      if (res.status_code == 200):
        with open(filename, "wb") as f:
          f.write(res.content)
          print('chapter '+localFold + '-' + str(imgCount) + ' down success')
          f.close()
      else:
        print('chapter '+localFold + '-' + str(imgCount) + ' down failed')
        # 针对一张情况，没有正确响应，认定没有接下来的图片了，就跳出循环
        # img_index = img_len
        # 知道总数的情况
        if (emptyRecord.count(imgCount) == 0):
          emptyImageList.append(imgCount)
    img_index += 1
  emptyRecord.extend(emptyImageList)
  with open(emptyFile, "wb") as f:
      resultStr = json.dumps(emptyRecord)
      # print(resultStr)
      f.write(resultStr.encode(encoding='UTF-8'))
      f.close()
  return

# 获取图片下载的地址，这种适用图片地址无法看出规律，
# 需要先请求 HTML网页，返回网页里面会包含图片得地址
# 每一张图片要请求一次 html，这种形式的要把数据先存放到本地
def get_every_img_address_html(localFold,downChapter):
  # 漫画网站本来应该有，但返回了空的数据
  emptyImageList = []
  img_list=[]
  fold_name = urllib.parse.unquote(localFold) + '/'
  imgListFile = baseRoot+fold_name+ imagesJsonFileName
  emptyFile = baseRoot+fold_name+ emptyJsonFileName
  totalFile = baseRoot+fold_name+ totalJsonFileName
  emptyRecord = []
  if (os.path.exists(emptyFile)):
    with open(emptyFile, "r",encoding="utf-8") as f:
        content = f.read()
        if (content):
          emptyRecord = json.loads(content)
  imageExistList = []
  if (os.path.exists(imgListFile)):
    with open(imgListFile, "r",encoding="utf-8") as f:
        content = f.read()
        if (content):
          imageExistList = json.loads(content)
  localTotalNum = 0
  if (os.path.exists(totalFile)):
    with open(totalFile, "r",encoding="utf-8") as f:
        content = f.read()
        if (content):
          localTotalNumObj = json.loads(content)
          localTotalNum = localTotalNumObj['total']
  page_num = len(imageExistList) + 1
  titleText = 'not found'

  while page_num <= maxPageNum:
  # while page_num <= 1:
    num_format = str(page_num)
    filename = reqList + downChapter + "_p"+ num_format + fileType
    res = requests.get(filename)

    if (res.status_code == 200):

      pattern = re.compile(r'<img class="img-fluid show-pic"+.*?\/>')
      result = pattern.findall(res.text)
      titlePattern = re.compile(r'<title>+.*?<\/title>')
      titleResult = titlePattern.findall(res.text)
      # 可以从页面中拿到总页数
      totalPattern = re.compile(r'data-total="+.*?"')
      totalResult = totalPattern.findall(res.text)
      # print('result',totalResult)
      # return

        # print('img_list',img_list)
      if (len(titleResult)):
        titleText = titleResult[0]
      if (len(totalResult)):
        totalCount = utils.getQuoteNumValue(totalResult[0])
        if(totalCount != localTotalNum):
          localTotalNum = totalCount
        if (totalCount != maxPageNum):
          global maxPageNum
          maxPageNum = totalCount
      if (len(result)):
        print('get chapter '+ localFold +' image '+str(page_num) + ' url success')
        imgSrcValue = utils.getImgSrc(result[0])
        img_list.append(imgSrcValue)
      else:
        # 避免多余的数据添加
        if (localTotalNum and page_num <= localTotalNum):
          print('get chapter '+ localFold +' image ' + str(page_num) + ' empty')
          # 放一个占位符，这样生成图片顺序就保持一致，也好知道那里出问题了
          img_list.append('')
          if (emptyRecord.count(page_num) == 0):
            emptyImageList.append(page_num)
      page_num += 1
    else:
      # 没有正确响应，认定没有接下来的图片了，就跳出循环，把获取成功的数据保存
      print('get chapter '+ localFold +' image ' + str(page_num) + ' error')
      page_num = maxPageNum
  # print(img_list)
  imageExistList.extend(img_list)
  emptyRecord.extend(emptyImageList)
  name = "title.md"
  dataFilename = baseRoot+fold_name+ name
  with open(dataFilename, "w+",encoding="utf-8") as f:
      f.write(titleText)
      f.close()

  with open(imgListFile, "wb") as f:
      resultStr = json.dumps(imageExistList)
      # print(resultStr)
      f.write(resultStr.encode(encoding='UTF-8'))
      f.close()
  with open(totalFile, "wb") as f:
      resultStr = json.dumps({'total':localTotalNum})
      # print(resultStr)
      f.write(resultStr.encode(encoding='UTF-8'))
      f.close()
  with open(emptyFile, "wb") as f:
      resultStr = json.dumps(emptyRecord)
      # print(resultStr)
      f.write(resultStr.encode(encoding='UTF-8'))
      f.close()
  # down_img(localFold,img_list,0)
  return


# 获取所有章节数据，并存放到本地
def getChaptersData():
  reqUrl = 'https://www..com/manhua/'+ comicMark +'/'
  res = requests.get(reqUrl)
  # print(res.status_code)
  if (res.status_code == 200):
    pattern = re.compile(r'<ol class="links-of-books num_div"+.*?>([\s\S]*?)</ol*?>')
    versionResult = pattern.findall(res.text)
    # 可能有多个版本，例如黑白和彩色，所以这里需要再处理一次
    pattern2 = re.compile(r'href="/manhua/'+comicMark+'/+.*?html"')
    result = pattern2.findall(versionResult[0])

    # 正传
    zhengZhuan = []
    # print(len(result))
    # print(result)
    if (len(result)):
      for index,value in enumerate(result):
        valueSplit1 = value.split('.')
        valueSplit2 = valueSplit1[0].split('/')
        valueSplit2Len = len(valueSplit2)
        chapterLink = valueSplit2[valueSplit2Len-1]
        zhengZhuan.append(chapterLink)


    # print(len(result))
      with open(chapterFile, "wb") as f:
        resultStr = json.dumps(zhengZhuan)
        f.write(resultStr.encode(encoding='UTF-8'))
        print('get chapter list success')
        f.close()
  else:
    print('get chapter failed')
  return

# 单独获取图片并存放到本地 json 文件
def getImagesData():
  with open(chapterFile, "r",encoding="utf-8") as f:
    content = f.read()
    chapterList = json.loads(content)
    chapterNum = len(chapterList)
    startDire = 1
    while startDire <= chapterNum:
      startDownChapter = chapterList[startDire-1]
      utils.createFold(baseRoot, str(startDire))
      get_every_img_address_html(str(startDire),str(startDownChapter))
      startDire += 1
  return

# 根据本地图片 json 文件区请求相应图片，并保存到本地
def downAllImages():
  with open(chapterFile, "r",encoding="utf-8") as f:
    content = f.read()
    chapterList = json.loads(content)
    chapterNum = len(chapterList)
    startDire = 1
    while startDire <= chapterNum:
      fold_name = urllib.parse.unquote(str(startDire)) + '/'
      imgListFile = baseRoot + fold_name + imagesJsonFileName
      with open(imgListFile, "r",encoding="utf-8") as f:
        imgContent = f.read()
        imgList = json.loads(imgContent)
        down_img(str(startDire),imgList,1)
      startDire += 1
  return

def main():
  getChaptersData()
  # getImagesData()
  # downAllImages()
  return


# 主程序
main()

print('all done')

File: script/get-comic.js

const fs = require('fs');
const {
  createFold,
  writeLocalFile,
  requestPromise,
  readJsonFile,
  removeRepeat,
} = require('./utils')
const {
  getChapterContainerReg,
  getChapterReg,
  classifyData,
  formatChapter,
  sortChapterLink,
  globalExpand,
  creatClassifyFold,
  getChapterImageData,
  getImageType,
  getImageHeader,
} = require('./helper')

const baseRoot = '../comic/diYuLe/'
const comicMark = '31737'
// const chapterReqUrl = "https://www.ykmh.com/manhua/" + comicMark + '/'
const site = 'http://www.wuqimh.com'
const chapterReqUrl = site + "/" + comicMark + '/'
const chapterFile = baseRoot + 'chapter.json'
const imagesJsonFileName = 'images.json'
const titleFileName = 'title.md'
const emptyJsonFileName = 'empty.json'
const baseEmptyJsonFile = baseRoot+'empty.json'
const baseFailJsonFile = baseRoot+'down-fail.json'

// 获取所有章节数据，并存放到本地
async function getChaptersData() {
  const result = await requestPromise(`${chapterReqUrl}`,{reqType:'http'})

  const linkReg = getChapterContainerReg(1)
  const matchResult = result.match(linkReg)
  // console.log('---matchResult---')
  // console.log(matchResult)

  if (matchResult && matchResult.length) {
    const chapterReg = getChapterReg(3,comicMark)
    // 默认先取第一个，一个结束后，手动修改
    const chapterLink = matchResult[0]
    const linkMatch = chapterLink.match(chapterReg)
    // console.log('---linkMatch---')
    // console.log(linkMatch)

    // return;

    if (linkMatch && linkMatch.length) {
      // 先按照 正式连载,短篇,卷附录，单行本, 进行分类，然后再分别放入数组
      const allLink = classifyData(linkMatch,4)
      // let allLink = linkMatch.reduce((acc,cur) => {
      //   acc.push(formatChapter(cur,1))
      //   return acc
      // },[])
    // console.log('---allLink---')
    // console.log(allLink)
      // 有的情况可以根据数字大小排序
      // allLink = sortChapterLink(allLink,1)
      writeLocalFile(chapterFile,JSON.stringify(allLink))
      console.log('get all chapter links success')
    }

  }


}

// 获取每个章节下所有图片的链接并存放到本地 json 文件
async function getImagesData(type) {
  // 有就用
  // globalExpand()
  const useWay2 = type == 2;
  const fileData = readJsonFile(chapterFile)

  creatClassifyFold(fileData,baseRoot,1)

  let baseFail = readJsonFile(baseEmptyJsonFile)
    // console.log('---chapterList---')
    // console.log(chapterList)
  let chapterList = fileData
  const classify = 'appendix' // 有 4 个值 serial short single appendix
  if (useWay2) {
    chapterList = fileData[classify]
  }
  const chapterNum = chapterList.length
  let startDire = 1 // 跟本地的文件夹命名顺序一致，从 1 开始
  while (startDire <= chapterNum) {
  // while (startDire <= 1) { // 测试用
    let startDownChapter = chapterList[startDire-1]
    let preBaseRoot = `${baseRoot}`
    if (useWay2) {
      startDownChapter = startDownChapter.link
      preBaseRoot = `${baseRoot}${classify}/`
    }
    createFold(`${preBaseRoot}${startDire}`)

    const pathPre = `${preBaseRoot}${startDire}/`
    const imagesFile = `${pathPre}${imagesJsonFileName}`
    const titleFile = `${pathPre}${titleFileName}`
    if (fs.existsSync(titleFile) || baseFail.includes(startDownChapter)) {
      // console.log(`chapter ${startDire} all image src exist`)
      startDire += 1
      continue;
    }
    let url = `${chapterReqUrl}${startDownChapter}`
    let imageData = {}
    let reqOptions = {}
    if (useWay2) {
      url = `${site}${startDownChapter}`
      reqOptions = {reqType:'http'}
    }
    const res = await requestPromise(url,reqOptions).catch((e) => {
      console.log(`chapter fail ${startDownChapter}`)
      baseFail.push(startDownChapter)
      writeLocalFile(baseEmptyJsonFile,JSON.stringify(baseFail))
    })

    if (!res) {
      startDire += 1
      continue;
    }

    imageData = getChapterImageData(res,2)
    if (!imageData.total) {
      return;
    }
    const imagesContent = JSON.stringify(imageData)

    await writeLocalFile(imagesFile,imagesContent)
    await writeLocalFile(titleFile,imageData.title)
    console.log(`get chapter ${startDire} all image src`)
    startDire += 1
  }
}

// 对获取失败的章节的处理,未完成
async function getDownFailImagesData(type) {
  const useWay2 = type == 2;
  const fileData = readJsonFile(baseFailJsonFile)

    // console.log('---chapterList---')
    // console.log(chapterList)
  const classify = 'appendix' // 有 4 个值 serial short single appendix
  chapterList = fileData[classify]
  const chapterNum = chapterList.length
  let startDire = 1 // 跟本地的文件夹命名顺序一致，从 1 开始
  while (startDire <= chapterNum) {
  // while (startDire <= 1) { // 测试用
    let startDownChapterObj = chapterList[startDire-1]
    let filePos = startDownChapterObj.file
    startDownChapterUrl = startDownChapterObj.link
    let preBaseRoot = `${baseRoot}${classify}/`

    const pathPre = `${preBaseRoot}${filePos}/`
    const imagesFile = `${pathPre}${imagesJsonFileName}`
    const titleFile = `${pathPre}${titleFileName}`
    // if (fs.existsSync(titleFile)) {
      // console.log(`chapter ${startDire} all image src exist`)
      // startDire += 1
      // continue;
    // }
    let url = `${chapterReqUrl}${startDownChapterUrl}`
    let imageData = {}
    let reqOptions = {}
    if (useWay2) {
      reqOptions = {reqType:'http'}
    }
    const res = await requestPromise(url,reqOptions).catch((e) => {
      console.log(`chapter fail ${filePos}`)
    })

    if (!res) {
      startDire += 1
      continue;
    }

    imageData = getChapterImageData(res,2)
    if (!imageData.total) {
      return;
    }
    const imagesContent = JSON.stringify(imageData)

    await writeLocalFile(imagesFile,imagesContent)
    await writeLocalFile(titleFile,imageData.title)
    console.log(`get chapter ${startDire} all image src`)
    startDire += 1
  }
}

async function downAllImages() {
  const chapterList = readJsonFile(chapterFile)
    // console.log('---chapterList---')
    // console.log(chapterList)
  const chapterNum = chapterList.length
  let startDire = 1
  while (startDire <= chapterNum) {
  // while (startDire <= 204) { // 测试用
    const pre = `${baseRoot}${startDire}/`
    const emptyJsonFile = `${pre}${emptyJsonFileName}`
    const imgListFile = `${pre}${imagesJsonFileName}`
    const imagesContent = readJsonFile(imgListFile)
    const {total,list} = imagesContent
    let failData = readJsonFile(emptyJsonFile)

    const dataLen = list.length
    for (let index = 0; index < dataLen;index++) {
      const url = list[index];
      // console.info('url',url)
      const imagType = getImageType(url,2)
      const imagePath = `${pre}${index+1}${imagType}`
      if (fs.existsSync(imagePath)) {
        console.log(`image exist ${imagePath}`)
        continue;
      }
      if (failData.includes(imagePath)) {
        continue;
      }
      const imgHeader = getImageHeader(2)
      const res = await requestPromise(url,{encoding:'binary',headers:imgHeader}).catch((e) =>{
        console.log(`down fail ${imagePath}`)
        failData.push(url)
        writeLocalFile(emptyJsonFile,JSON.stringify(failData))
      })

      if (!res) {
        continue;
      }

      await writeLocalFile(imagePath,res,'binary')
      console.log(`down success ${imagePath}`)
    }
    startDire += 1
  }
}

async function test () {

  const url = 'http://www.mangabz.com/m19561/chapterimage.ashx?cid=19561&page=1&key=&_cid=19561&_mid=280&_dt=2021-03-12+203A21%3A27%3A30&_sign=56a8ed542ee44d45b542a249b7683d68'
  // const url = 'http://www.mangabz.com/m19561/'
  const res = await requestPromise(`${url}`,{reqType:'http',headers:{}})
  const result = eval(res)
  console.log('---res---')
  console.log(res)


}

// getChaptersData()
getImagesData(2)
// downAllImages()
// test()
